File: ksudoku.pl
% :- type file        ==  atom.

% :- type sspec ---> s(size, board).
% :- type size  == int.
% :- type field == list(info).
% :- type info ---> e; o; s; w; v(int).
% :- type board == list(list(field)).

% :- type ssol == list(list(int)).

% sudoku(SSpec, SSol):
% SSol az SSpec feladványt kielégítõ megoldás.
% :- pred sudoku(sspec::in, ssol::out).

:- module(sudoku_keret,
	  [
	   sudoku_be/2,		% sudoku_be(file::in, sspec::out)
	   sudoku_ki/2,		% sudoku_ki(file::in, sspec::in)
	   megold/2,		% megold(file::in, file::in)
	   stopper/2,		% stopper(file::in, file::in)
           teljes_teszt/1       % teljes_teszt(integer::in)
	  ]
	 ).

:- use_module(library(lists), [nth1/3,select/3]).
:- use_module(library(between), [between/3]).

%-------------- Bejaratok ------------------------------------

% sudoku_be(+File, -Feladvany):
% A File-bol beolvashato a Feladvany altal megadott feladat.
sudoku_be(File, Feladvany) :-
	open_file(File, read, Stream),
	call_cleanup(read_in(Stream, Characters),
		     close(Stream)),
	parse_data(Characters, Feladvany).

% sudoku_ki(+File, +Feladvany):
% A Feladvany feladat megoldasat kiirja a File allomanyba.
sudoku_ki(File, Feladvany) :-
	megoldasok(Feladvany, Megoldasok),
	open_file(File, write, Stream),
	current_output(OOut),
	set_output(Stream),
	call_cleanup(print_solutions(Megoldasok),
		     (set_output(OOut),close(Stream))).

% megoldasok(+Feladvany, -Megoldasok):
% A Feladvany feladat osszes megoldasa
% a Megoldasok rendezett lista.
megoldasok(Feladvany, Megoldasok) :-
	findall(Megoldas,
		user:sudoku(Feladvany, Megoldas),
		Megoldasok0),
	sort(Megoldasok0, Megoldasok).

% megold(+FileIn, +FileOut):
% A FileIn allomanybol beolvasott feladat minden megoldasat
% kiirja a FileOut allomanyba.
megold(FileIn, FileOut) :-
	sudoku_be(FileIn, Feladvany),
	ensure_loaded(user:sudoku),
	megoldasok(Feladvany, Megoldasok),
	open_file(FileOut, write, Stream),
	current_output(OOut),
	set_output(Stream),
	call_cleanup(print_solutions(Megoldasok),
		     (set_output(OOut),close(Stream))).

% stopper(+FileIn, +FileOut):
% Mint megold/2, de a futashoz szukseges idot is kiirja.
stopper(FileIn, FileOut) :-
	stopper(FileIn, FileOut, []).

% stopper(+FileIn, +FileOut, +Options): Mint stopper/2 de elfogad egy
% Options opciolistat.  Ha az opciolista eleme sols(Sols), akkor a
% megoldaslistat egyesiti Sols-sal. Ha az opciolista eleme nonl, akkor a
% a kiiras utan nem emel sort.
stopper(FileIn, FileOut, Options) :-
	sudoku_be(FileIn, Feladvany),
	ensure_loaded(user:sudoku),
	statistics(runtime, [T0|_]),
	megoldasok(Feladvany, Megoldasok),
	statistics(runtime, [T1|_]),
	T is T1 - T0,
	length(Megoldasok, H),
	format('Feladvány: ~a, megoldásszám:~|~t~d~4+, futási idõ:~|~t~3d~8+ s', [FileIn,H,T]),
	(   memberchk(nonl, Options) -> true
	;   nl
	),
	(   memberchk(sols(Sols), Options) -> Sols = Megoldasok
	;   true
	),
	open_file(FileOut, write, Stream),
	current_output(OOut),
	set_output(Stream),
	call_cleanup(print_solutions(Megoldasok),
		     (set_output(OOut),close(Stream))).

%-------------- Allomanyok kezelese --------------------------

% open_file(+File, +Mode, -Stream):
% A File allomanyt megnyitja a Stream folyam formajaban Mode modban.
open_file(user, M, S) :-
	!, standard_stream(M, S).
open_file(File, Mode, Stream) :-
	open(File, Mode, Stream).

% standard_stream(?Mode, ?Stream):
% A Mode modu szabvanyos folyam a Stream.
standard_stream(read, user_input).
standard_stream(write, user_output).

%-------------- Beolvasas ------------------------------------

% read_in(+S, -L):
% S folyambol beolvashato karakterek listaja L.
read_in(S, L) :-
	get_code(S, C),
	read_in(C, S, L).

% read_in(+C, +S, -L):
% Ha C az S folyambol elozoleg beolvasott karakter, akkor L az S-bol
% beolvashato maradek karakterek listaja, C-t is beleertve.
% Ha C = -1, akkor L = [].
read_in(C, _, L) :-
	C =:= -1, !, L = [].
read_in(C, S, [C|L]) :-
	get_code(S, C1),
	read_in(C1, S, L).

% parse_data(+Chars, -Feladvany):
% Chars fuzerbol kielemezheto egy Feladvany feladvanyleiro
parse_data(Chars, Feladvany) :-
	descriptor(Feladvany, Chars, Rest),
	correct_syntax(Rest),
	!.
parse_data(_, _) :-
	warning('A bemeneti allomany nem megfelelo formatumu.',[]),
	fail.

% correct_syntax(+Rest):
% A Rest fuzer a bemenet helyes maradeka, azaz ures.
correct_syntax(Rest) :-
	Rest = [_|_], !,
	chars_of_type(mid_line, L, Rest, _),
	warning('Nem vart karakter(ek): "~s"', [L]),
	fail.
correct_syntax(_).

%-------------- DCG elemzes ----------------------------------

% descriptor(SSpec): Kielemezheto egy SSpec feladvany-leiro
descriptor(s(K,Board)) -->
	spaces,
	nat_number(K),
	{N is K*K},
	newline,
	board_data(N, N, Board),
	chars_of_type(layout, _).

% Kielemezheto egy (tizes szamrendszerbeli) Int termeszetes szam.
nat_number(Int) -->
	chars_of_type(digit, Ds),
	{Ds = [_|_], number_codes(Int, Ds)}.

% board_data(Y, N, Board): Kielemezheto egy N*N-es feladvany Y sorbol
% allo Board resztablazata.
board_data(0, _, []) --> [].
board_data(Y, N, [R|Rs]) -->
	board_row(N, N, R), 
	chars_of_type(layout, _),
	{Y1 is Y-1},
	board_data(Y1, N, Rs).

% board_row(X, N, R): Kielemezheto egy N*N-es feladvany reszekent
% egy X elembol allo R sor-reszlet.
board_row(X, N, [F|Fs]) -->
	spaces,
	field(N, F),
	{X1 is X-1},
	board_row(X1, N, Fs).
board_row(0, _, []) --> [].

% field(N, F): Kielemezheto egy N*N-es feladvany F tablazat-mezoje.
field(_, []) -->
	"-" .
field(N, Is) -->
	{ Is = [_|_] },
	infos("vsweo", N, Is).

% infos(Ts, N, Is) Kielemezheto egy N*N-es feladvanybeli infok
% Ts tipusu Is listaja.
infos(Ts0, N, [I|Is]) -->
	info(Ts0, N, I, Ts1), !,
	infos(Ts1, N, Is).
infos(_, _, []) --> [].

% info(Ts0, N, I, Ts): Kielemezheto egy N*N-es feladvanybeli, a Ts0 tipusok 
% egyikebe tartozo I info, Ts a meg lehetseges tipusok listaja.
info(Ts0, N, I, Ts) -->
	{select(T, Ts0, Ts)},
	(   {T = 0'v}, nat_number(J),
	    {   J >= 1, J =< N -> I = v(J)
	    ;   warning('A ~w szaminfo hibas: 1 =< ~w  =< ~w nem teljesul', [J,J,N]), fail
	    }
	;   [T], {atom_codes(I, [T])}
	), !.
	
% Kielemezheto egy ujsor.
newline -->
	spaces,
	char_of_type(vertical, _).

% Kielemezheto vizszintes iranyu nem lathato karakterek
% egy sorozata.
spaces -->
	chars_of_type(horizontal, _).

% layout_char(Dir, C):
% horizontal_layout_char(C): C  egy vizszintes iranyu
% nem lathato karakter.
horizontal_layout_char(0' ).
horizontal_layout_char(0'\t).
horizontal_layout_char(0'\r).

% vertical_layout_char(C): C  egy fuggoleges iranyu
% nem lathato karakter.
vertical_layout_char(0'\n).
vertical_layout_char(0'\f).
vertical_layout_char(0'\v).

% Kielemezheto egy Type tipusu C karakter.
char_of_type(Type, C) -->
	[C], {char_type(Type, C)}.

% chars_of_type(Type, Cs) -->
% Kielemezheto Type tipusu karakterek egy Cs listaja.	
chars_of_type(Type, [C|Cs]) -->
	char_of_type(Type, C), !,
	chars_of_type(Type, Cs).
chars_of_type(_, []) --> [].

% char_type(Type, C) :
% A C karakter Type tipusu.
char_type(layout, C) :-
	(   vertical_layout_char(C) -> true
	;   horizontal_layout_char(C)
	).
char_type(digit, D) :-
	D >= 0'0, D =< 0'9.
% Csak kisbetuket fogadunk el.
char_type(alpha, A) :-
	A >= 0'a, A =< 0'z.
char_type(mid_line, C) :-
	\+ vertical_layout_char(C).
char_type(horizontal, C) :-
	horizontal_layout_char(C).
char_type(vertical, C) :-
	vertical_layout_char(C).

%-------------- Kiiratas -------------------------------------

% print_solutions(+Megoldasok):
% A Megoldasok megoldaslista minden elemet
% kiirja az aktualis kimenetre.
print_solutions(Megoldasok) :-
	member(Megoldas, Megoldasok),
	write('\n-----'),
	print_board(Megoldas),
	fail.
print_solutions(_) :-
	write('\n-----\n').

% print_board(Tabla)
% Kiirja a Tabla tablat, ami lehet kitoltetlen is
print_board(B) :-
	length(B,H),
	K is integer(round(sqrt(H))),
	B = [R1|_],
	length(R1,W),
	between(1,H,Y),
	nth1(Y,B,R),
	nl,
	((Y mod K =:= 1 ; K == 1) -> nl ; true),
	between(1,W,X),
	nth1(X,R,F),
	Tab is 1+floor(log(10,K*K)+1)-floor(log(10,max(1,F))+1),
	(   between(1, Tab, _), write(' '), fail  % Tab db. szóköz kiírása
	;   true
	),
	(X > 1, X mod K =:= 1 -> write(' ') ; true),	
	write(F),
	fail.
print_board(_) :- nl.

%-------------- Seged-eljaras ------------------------------

% warning(+Txt, +Args):
% Kiirja a Txt szoveggel es Args argumentumlistaval
% megadott figyelmezteto szoveget.
warning(Txt, Args) :-
	print_message(warning, format(Txt, Args)).

%-------------- ISO kompatibilitas ---------------------------

% Az alabbi cel az ISO Prolog szabvanynak meg nem megfelelo, de
% DEC10 kompatiblis Prolog rendszerekben valo futtatashoz szukseges.

:- (   predicate_property(get_code(_,_), _) -> true
   ;   assert((get_code(S, C) :- get0(S, C))),
       assert((put_code(C) :- put(C))),
       assert((number_codes(N, C) :- number_chars(N, C)))
   ).

%-------------- experimental ---------------------------
% Az alábbi eljárást Eisenberger András ültette át Prologra Erlangból, és egészítette ki a megoldás ellenörzésével.

:- use_module(library(file_systems)).
:- use_module(library(timeout)).
:- use_module(library(samsort)).

% A 'tests' könyvtárban levõ összes "testXXXd.txt" tesztállomány esetén
%  - lefuttatja a tesztet Timeout másodperces idõkorláttal,
%  - ellenõrzi, hogy a testXXXs.txt állományban megadott megoldáshalmazt kapta,
%  - olvasható formában (lásd megold/2) kiírja az eredményt a 'tests_out_pl'
%    könyvtár testXXXt.txt nevû állományába.
% Az állománynevekben az XXX szám tetszõleges hosszúságú lehet.
teljes_teszt(Timeout) :-
	Time_MS is Timeout * 1000,
	(   directory_exists(tests_out_pl) -> true
	;   make_directory(tests_out_pl)
	),
	file_members_of_directory('tests', 'test*d.txt', FoundFiles),
	(   member(BaseName-_AbsPath, FoundFiles),
	    atom_concat('tests/', BaseName, InPath),
	    atom_concat(TestName, 'd.txt', BaseName),
	    atom_concat('tests/', TestName, 's.txt', SolsPath),
	    atom_concat('tests_out_pl/', TestName, 't.txt', OutPath),
            time_out(stopper(InPath, OutPath, [nonl,sols(Sols)]),
		     Time_MS, Result),
            (   Result == success ->
		samsort(Sols, SolsSorted),
		catch(read_term_from_file(SolsPath, SolsRead), _,
		      SolsRead = none),
		(   SolsRead == none -> write(' NINCS MEGOLDÁS\n')
		;   SolsRead == SolsSorted -> write(' HELYES\n')
		;   write(' NEM HELYES\n')
		)
	    ;	format('Feladvány: ~a, túllépte az idõkorlátot (~d s)~n',
		       [InPath,Timeout])
            ),
	    fail
	;   true
	).

% atom_concat(A, B, C, ABC): Az A, B es C atomok összefûzése ABC.
atom_concat(A, B, C, ABC) :-
	atom_concat(A, B, AB),
	atom_concat(AB, C, ABC).

% read_term_from_file(File, Term): A Term Prolog kifejezést beolvassa a
% File állományból.
read_term_from_file(File, Term) :-
	open_file(File, read, Stream),
	call_cleanup(read(Stream, Term),
		     close(Stream)).

File: ksudoku.erl
-module(ksudoku).
-author('hanak@inf.bme.hu, kapolnai@iit.bme.hu, modified by: Andras Eisenberger, Balazs Galambosi').
-vsn('$LastChangedDate: 2016-09-15 16:40:04 +0200 (cs, 15 szept 2016) $$').

% -compile(export_all).
-export([sudoku_be/1,sudoku_ki/2,megold/2,stopper/2,stopper/3, help/0,teljes_teszt/1,teszt/2]).

-import(sudoku,[sudoku/1]).

%% @type sspec() = {size(), board()}.
%% @type size()  = integer().
%% @type field() = [info()].
%% @type info()  = e | o | s | w | integer().
%% @type board() = [[field()]].

%% @type ssol() = [[integer()]].

%% @spec sudoku:sudoku(SSpec::sspec()) -> SSols::[ssol()].
%% @doc  SSols az SSpec feladványt kielégítõ megoldások listája.

%% ksudoku-specifikációk

%% @spec sudoku_be(FileIn::string()) -> SSpec::sspec().
%% @doc  SSpec a FileIn szövegfájlból beolvasott Sudoku-feladvány.
%%
sudoku_be(FileIn) ->
    case file:open(FileIn,read) of
        {ok, InputFile}    -> ok;
        {error, InputFile=R} -> throw({error,R,file:format_error(R)})
    end,
    {N,_} = case io:get_line(InputFile,'') of
		eof ->
		    throw({error,ures,lists:flatten(io_lib:format("Ures a(z) ~s fajl",[FileIn]))});
		I ->
		    string:to_integer(cleaned(I))
	     end,
    KT = tabla_be(InputFile,N),
    file:close(InputFile),
    case helyes_feladvany(KT) of
	true ->
	    KT;
	false ->
	    throw({error,hibas,lists:flatten(io_lib:format("Hibas a feladvany a(z) ~s fajlban",[FileIn]))})
    end.

tabla_be(InputFile,M) ->
    {M,tabla_be(InputFile,M*M,[])}.

tabla_be(_InputFile,0,Zss) ->
    lists:reverse(Zss);
tabla_be(InputFile,Meret,Zss) ->
    case io:get_line(InputFile,'') of
	eof ->
	    throw({error,keves,lists:flatten(io_lib:format("Tul keves sor van a input fajlban",[]))});
	Sor ->
	    Ss = string:tokens(cleaned(Sor)," "),
	    Zs = lists:map(fun str2infolist/1, Ss),
	    case Zs of
		[] ->
		    tabla_be(InputFile,Meret,Zss);
		_ ->
		    tabla_be(InputFile,Meret-1,[Zs|Zss])
	    end
    end.

cleaned(Sor) ->
    S0 = lists:filter(fun(C) -> C /= $\t end,Sor),
    string:strip(string:strip(S0),right,$\n).

str2infolist([$-]) ->
    [];
str2infolist(XXs=[X|_Xs]) when $0=<X andalso X=<$9 ->
    {I,Rs} = string:to_integer(XXs),
    [I|str2infolist(Rs)];
str2infolist([X|Xs]) ->
    [list_to_atom([X])|str2infolist(Xs)];
str2infolist([]) ->
    [].

%% @spec helyes_feladvany(SSpec::sspec()) -> B::bool().
%% @doc  B igaz, ha a feladvány megfelel a specifikációnak.
%%
helyes_feladvany({K,T}) ->
    KK = K*K,
    % a cellaméret jó
    (1 =< K) andalso (K =< 6) andalso
    % a sorok és soronként az oszlopok száma jó 
    (KK == length(T)) andalso
	(lists:all(fun(L) -> length(L) == KK end,T)) andalso
    % a tábla megfelel a sudoku-feladvány specifikációjának
%    lists:all(fun helyes_mezo/2,lists:concat(T)).
    lists:all(fun(Fs) -> helyes_mezo(KK,Fs) end,lists:concat(T)).

helyes_mezo(KK,Fs) ->
    length(lists:filter(fun erlang:is_integer/1,Fs)) < 2 andalso
        lists:all(fun(A) ->
                          length([1 || F <- Fs, F =:= A]) < 2
                  end, [s,w,e,o]) andalso
        lists:all(fun(F) ->
                          is_integer(F) andalso 1=<F andalso F=<KK
                              orelse lists:member(F, [s,w,e,o])
                              
                  end, Fs).

%% @spec ksudoku:sudoku_ki(FileOut::string(),SSols::[ssol()]) -> void()
%% @doc  Az SSols listában átadott Sudoku-megoldásokat kiírja
%%       a FileOut szövegfájlba.
%%
sudoku_ki(FileOut,SSols) ->
    case file:open(FileOut,write) of
        {ok, OutputFile}      -> ok;
        {error, OutputFile=R} -> throw({error,R,file:format_error(R)})
    end,
    tablak_ki(OutputFile,SSols),
    file:close(OutputFile).

tablak_ki(OutputFile,[SSol]) ->
    io:fwrite(OutputFile,"~s~n",[tabla_ki(SSol)]);
tablak_ki(OutputFile,[SSol|SSols]) ->
    io:fwrite(OutputFile,"~s~n~n",[tabla_ki(SSol)]),
    tablak_ki(OutputFile,SSols);
tablak_ki(OutputFile,[]) ->
    io:fwrite(OutputFile,"",[]).

tabla_ki([L]) ->
    sor_ki(L);
tabla_ki([L|Ls]) ->
    sor_ki(L) ++ "\n" ++ tabla_ki(Ls);
tabla_ki(_) ->
    [].

sor_ki([H|T]) ->
    I2L = fun(Int) -> io_lib:format(if length([H|T]) > 9 -> "~2w" ; true -> "~w" end, [Int]) end,
    I2L(H) ++ lists:foldr(fun(A,B) -> " " ++ I2L(A) ++ B end,"",T);
sor_ki([]) ->
    "".
%% @spec sudokusol_be(FileIn::string()) -> SSols::[ssol()].
%% @doc  SSpec a FileIn szövegfájlból beolvasott Sudoku-feladvány.
%%       Készítette: Eisenberger András.
sudokusol_be(FileSol) ->
    case file:open(FileSol,read) of
        {ok, InputFile}      -> ok;
        {error, InputFile=R} -> throw({error,R,file:format_error(R)})
    end,
    case io:read(InputFile, "") of
        {ok, Sol}       -> ok;
        {error, Sol=R2} -> throw({error,R2,file:format_error(R2)})
    end,
    file:close(InputFile),
    Sol.

%% @spec ksudoku:megold(FileIn::string(),FileOut::string()) -> void()
%% @doc  Beolvas egy feladványt a FileIn szövegfájlból és összes
%%       megoldását kiírja a FileOut szövegfájlba. Ehhez
%%       felhasználja a sudoku:sudoku/1 függvényt.
%%
megold(FileIn,FileOut) ->
    KT = sudoku_be(FileIn),
    SSols = sudoku:sudoku(KT), 
    sudoku_ki(FileOut,SSols).

%% @spec ksudoku:stopper(FileIn::string(),FileOut::string()) -> void()
%% @doc  Mint ksudoku:megold/2, de a végén kiírja a FileIn nevet,
%%       a megoldások számát és a futási idõt is.
stopper(FileIn,FileOut) -> stopper(FileIn,FileOut,"").

%% @spec ksudoku:stopper(FileIn::string(),FileOut::string(),FileSol::string()) -> void()
%% @doc  Mint ksudoku:megold/2, de a végén kiírja a FileIn nevet,
%%       a megoldások számát és a futási idõt is. Ha lehet, összeveti FileSol megoldással is.
stopper(FileIn,FileOut,FileSol) ->
    KT = sudoku_be(FileIn),
    erlang:garbage_collect(),timer:sleep(10),
    _T1 = statistics(runtime),
    SSols = sudoku:sudoku(KT),
    {_,T2} = statistics(runtime),
    sudoku_ki(FileOut,SSols),
    io:format("Feladvany: ~s, ", [FileIn]),
    try   % Eisenberger András kiegészítése
        case lists:sort(sudokusol_be(FileSol)) =:= lists:sort(SSols) of
            true  -> io:format("HELYES ");
            false -> io:format("ROSSZ ")
        end
    catch
        {error, _, _} -> error%io:format("NINCS MO ")
    end,
    io:format("megoldas: ~3w, futasi ido: ",
	      [length(SSols)]),
    if T2 >= 10000 -> io:format("~5w s\n", [ T2 div 1000 ]);
%       T2 >= 1000 -> io:format("~5.2f s\n", [ T2 / 1000 ]);
       true        -> io:format("~5w ms\n", [ T2 ])
    end.


%% ------- experimental -----


help() ->
    io:format(
      "teljes_teszt(Timeout): meg kiserleti stadiumban van."
      "Megoldja az osszes, tests/testXXXd.txt specifikaciot idolimittel.~n"
      "Pelda: futtatas bash-bol, 10s idolimit, megoldasok tests/testXXXt.txt fajlokba: ~n"
      "$ erl -noshell -eval 'ksudoku:teljes_teszt(10).' -s init stop~n"
     ).

%% @spec teljes_teszt(Timeout::integer()) -> [done|timeout].
%% @doc A 'tests' könyvtárban levõ összes "testXXXd.txt" tesztállomány esetén
%%  - lefuttatja a tesztet Timeout másodperces idõkorláttal,
%%  - ellenõrzi, hogy a testXXXs.txt állományban megadott megoldáshalmazt kapta,
%%  - olvasható formában (lásd megold/2) kiírja az eredményt a 'tests_out_er'
%%    könyvtár testXXXt.txt nevû állományába.
%% Az állománynevekben az XXX szám tetszõleges hosszúságú lehet.
teljes_teszt(Timeout) ->
    Dir = "tests/",
    ODir = "tests_out_er/",
    case file:make_dir(ODir) of
        ok              -> ok;
        {error, eexist} -> ok;
        {error, Rd}     -> throw({error,Rd,file:format_error(Rd)})
    end,
    case file:list_dir(Dir) of
        {error, R} ->
            throw({error,R,file:format_error(R)});
        {ok, Content} ->
            IOFiles = [{Dir++Input,ODir++Output,Dir++Solution}
                       || Input <- lists:sort(Content),
                          {match,_} <- [re:run(Input, "test[0-9]*d\.txt")],
                          Solution <- [re:replace(Input, "d\.txt\$", "s.txt", [{return, list}])],
                          Output <- [re:replace(Input, "d\.txt\$", "t.txt", [{return, list}])]
                      ],
            teszt(Timeout, IOFiles)
    end.

%% @spec teszt(Timeout::integer(),
%%            IOFiles::[{Input::string(),Output::string(),Solution::string()}]) -> [done|timeout].
%% @doc Megoldja az összes Input fájlban található specifikációt Timeout idõlimittel,
%%      megoldásokat az Output fájlokba teszi. Ha lehet, összeveti a megoldásokat Solution-nel.
teszt(Timeout, IOFiles) ->
    [begin
         process_flag(trap_exit, true),
         Worker = spawn_link(fun() -> stopper(Input, Output, Solution) end),
         Timer  = spawn_link(fun() -> timer:sleep(Timeout * 1000) end),
         receive
             {'EXIT', Worker, _Reason} -> 
                 exit(Timer, kill),
                 receive {'EXIT', Timer, killed} -> ok end, % feldolgozzuk a halalat
                 done;
             {'EXIT', Timer, _Reason} ->
                 exit(Worker, kill),
                 io:format("Feladvany: ~s -- Timeout (~p s)~n", [Input,Timeout]),
                 receive {'EXIT', Worker, killed} -> ok end, % feldolgozzuk a halalat
                 timeout
         end
     end
     || {Input, Output, Solution} <- IOFiles
    ].
